TextToSpeech/impl/TTSSpeaker.cpp

```diff
@@ -0,0 +1,847 @@
+    // m_flushed (as nothing is being spoken, which needs bail out)
+    if(state == false)
+        m_flushed = false;
+}
+
+void TTSSpeaker::queueData(SpeechData data) {
+    std::lock_guard<std::mutex> lock(m_queueMutex);
+    m_queue.push_back(data);
+    m_condition.notify_one();
+}
+
+void TTSSpeaker::flushQueue() {
+    std::lock_guard<std::mutex> lock(m_queueMutex);
+    m_queue.clear();
+}
+
+SpeechData TTSSpeaker::dequeueData() {
+    std::lock_guard<std::mutex> lock(m_queueMutex);
+    SpeechData d;
+    d = m_queue.front();
+    m_queue.pop_front();
+    m_flushed = false;
+    return d;
+}
+
+bool TTSSpeaker::waitForStatus(GstState expected_state, uint32_t timeout_ms) {
+    // wait for the pipeline to get to pause so we know we have the audio device
+    if(m_pipeline) {
+        GstState state;
+        GstState pending;
+
+        auto timeout = std::chrono::system_clock::now() + std::chrono::milliseconds(timeout_ms);
+
+        do {
+            std::unique_lock<std::mutex> mlock(m_queueMutex);
+            m_condition.wait_until(mlock, timeout, [this, &state, &pending, expected_state] () {
+                    // Speaker has flushed the data, no need wait for the completion
+                    // must break and reset the pipeline
+                    if(m_flushed) {
+                        TTSLOG_VERBOSE("Bailing out because of forced text queue (m_flushed=true)");
+                        return true;
+                    }
+
+                    gst_element_get_state(m_pipeline, &state, &pending, GST_CLOCK_TIME_NONE);
+                    if(state == expected_state)
+                        return true;
+
+                    return false;
+                });
+        } while(!m_flushed && state != expected_state && timeout > std::chrono::system_clock::now());
+
+        if(state == expected_state) {
+            TTSLOG_VERBOSE("Got Status : expected_state = %d, new_state = %d", expected_state, state);
+            return true;
+        }
+
+        TTSLOG_WARNING("Timed Out waiting for state %s, currentState %s",
+                gst_element_state_get_name(expected_state), gst_element_state_get_name(state));
+        return false;
+    }
+
+    return true;
+}
+
+// GStreamer Releated members
+void TTSSpeaker::createPipeline() {
+    m_isEOS = false;
+
+    if(!m_ensurePipeline || m_pipeline) {
+        TTSLOG_WARNING("Skipping Pipeline creation");
+        return;
+    }
+
+    TTSLOG_WARNING("Creating Pipeline...");
+    m_pipeline = gst_pipeline_new(NULL);
+    if (!m_pipeline) {
+        m_pipelineConstructionFailures++;
+        TTSLOG_ERROR("Failed to create gstreamer pipeline");
+        return;
+    }
+
+    m_source = gst_element_factory_make("souphttpsrc", NULL);
+
+    // create soc specific elements
+#if defined(BCM_NEXUS)
+    GstElement *decodebin = NULL;
+    decodebin = gst_element_factory_make("brcmmp3decoder", NULL);
+    m_audioSink = gst_element_factory_make("brcmpcmsink", NULL);
+#endif
```

Comment:
I think this is against VAP requirements, i.e. we shouldn't use SoC specific plugins directly. has to be addressed sooner or later

Reply:
Hi Eugene, I have a doubt. Should rdk services as well follow VAP?

Reply:
Good question, I think it makes sense for services as well. I mean it should just work in a generic way.

Reply:
We will remove the explicit SOC specific plugins as part of RDK-20988.